packages/account/src/providers/provider.test.ts

```diff
@@ -1045,4 +1038,66 @@ describe('Provider', () => {
 
     await Promise.all(promises);
   });
+
+  test('requestMiddleware modifies the request before being sent to the node [sync]', async () => {
+    const fetchSpy = vi.spyOn(global, 'fetch');
+    await Provider.create(FUEL_NETWORK_URL, {
+      requestMiddleware: (request) => {
+        request.headers ??= {};
+        (request.headers as Record<string, string>)['x-custom-header'] = 'custom-value';
+        return request;
+      },
+    });
+
+    const requestObject = fetchSpy.mock.calls[0][1];
+
+    expect(requestObject?.headers).toMatchObject({
+      'x-custom-header': 'custom-value',
+    });
+  });
+
+  test('requestMiddleware modifies the request before being sent to the node [async]', async () => {
+    const fetchSpy = vi.spyOn(global, 'fetch');
+    await Provider.create(FUEL_NETWORK_URL, {
+      requestMiddleware: (request) => {
+        request.headers ??= {};
+        (request.headers as Record<string, string>)['x-custom-header'] = 'custom-value';
+        return Promise.resolve(request);
+      },
```

Comment:
Was the async variation specifically requested by someone? Trying to understand the need for it / where it could be used

Reply:
Somebody might want to fetch something from some other API to include in every call:
```ts
      requestMiddleware: (request) => {
        const credentials = await fetchSomeOtherCredentials();
        request.headers ??= {};
        (request.headers as Record<string, string>)['x-my-custom-credential'] = credentials.value;
        return Promise.resolve(request);
      }
```

Reply:
I took inspiration for the async variation from `GraphQLRequest` which has the same `requestMiddleware` option that can be passed into it. I couldn't just use that library's functionalities, though, because I had to cover subscriptions, for which we have a custom implementation. This will have to be the case for each functionality we extend `fetch` with: we can never benefit from the various `graphql-request` features because we'll always have to roll out custom logic to cover subscriptions.